Option: http error retries

dadah89 opened the pull request:
### Describe
In the case that we're downloading thousands of files at once, if `OpenURI::HTTPError` gets raised after 2 retries, the sync is aborted, and the next run will start over from the beginning. Hopefully increasing the number of retries will prevent sync failures, if not I'll take a look into adding support for partial downloads.

Example:
```
bin/s3-meta-sync --retries 10 -V my_s3_bucket:translations --no-local-changes tmp/translations
Downloading translations/.s3-meta-sync
```

Other changes:
 * prefix temp folder with s3ms
 * ensure temp folder is removed
 * update tests to run on ruby 2.4.1.

Ran tests on ruby 2.4.1 and 2.2.3:

```
(2.4.1) vagrant: /harmony/code/s3_meta_sync (master)$ rspec
.........................................................

Finished in 1 minute 36.08 seconds (files took 0.38627 seconds to load)
57 examples, 0 failures
```

```
(2.2.3) vagrant: /harmony/code/s3_meta_sync (igor/retry_remove_tmp_dir)$ rspec
.........................................................

Finished in 1 minute 37.91 seconds (files took 0.63786 seconds to load)
57 examples, 0 failures
```

### Reviewers
/cc @grosser @zendesk/i18n-dev

### Risks
 * Low. Extra options and small refactoring.

dadah89 commented on spec/s3_meta_sync_spec.rb:
```diff
@@ -21,7 +21,13 @@ def sh(command, options={})
   let(:bucket) { config[:bucket] }
-  let(:foo_md5) { "---\n:files:\n  xxx: 0976fb571ada412514fe67273780c510\n" }
+  let(:foo_md5) do |variable|
+    if RUBY_VERSION < '2.4.0'
```
`to_yaml` behaves slightly differently on ruby 2.4

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -83,6 +83,18 @@ def download(source, destination)
       end
     end
 
+    def mktmpdir
```
this looks sketchy ... better use regular mktmpdir and then handle issues in the ensure if they occur instead of rolling our own

dadah89 replied:
I'm using the syntax suggested here https://ruby-doc.org/stdlib-2.3.1/libdoc/tmpdir/rdoc/Dir.html (last comment block)
would renaming the method look better?

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -244,22 +256,27 @@ def download_content(path)
+    rescue OpenURI::HTTPError, Errno::ECONNRESET => e
+      max_retries = @config[:max_retries] || 2
+      retries ||= Hash.new(0)
+      retries[e.class] += 1
```
don't think we need that ... just regular 10 retries should be fine ... usually always  single issue ... also this is per file downloaded and not a global var

dadah89 replied:
works for me, will update

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -83,6 +83,18 @@ def download(source, destination)
+        # Sometimes remove_entry fails, trying really hard:
+        log "rm -rf #{dir}: #{`rm -rf #{dir}`.inspect}"
+      end
+    end
```
Prefer to keep mktmpdir and only do cleanup and log if something went wrong ... we should trust ruby to work by default ...

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -210,12 +222,12 @@ def download_meta(destination)
-      if retries == 2
-        raise RemoteWithoutMeta
-      else
+      if retries <= 1
```
never decrements ... infinite loop ?

dadah89 replied:
`retries` starts at 1, gets incremented in the next line, and if there is another exception, it'll hit the else condition

grosser replied:
ah yes! ... so `retries == 1` would also work ? :D

dadah89 replied:
that is inconsistent with the other retry block though, I'll move the increment outside the if/else block

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -244,22 +256,27 @@ def download_content(path)
+      http_error_retries ||= 0
+      http_error_retries += 1
+      if http_error_retries <= max_retries
+        log "#{e.class} error downloading #{path}, retrying (at #{http_error_retries})"
```
logging here is nice ... should show early that something is going wrong

grosser commented on spec/s3_meta_sync_spec.rb:
```diff
@@ -505,7 +548,11 @@ def sync(command, options={})
         expect(result).to eq <<-TXT.gsub(/^ {10}/, "").strip
           Downloading bar/.s3-meta-sync
+          OpenURI::HTTPError error downloading bar/.s3-meta-sync, retrying (at 1)
+          OpenURI::HTTPError error downloading bar/.s3-meta-sync, retrying (at 2)
```
that "at 2" looks strange ... how about "Retrying 3/4" ?

grosser commented:
even if it's suggested in the docs,I think it's still better to rely on the
automatic cleanup

On Wed, Aug 9, 2017 at 2:57 PM, Igor Justino <notifications@github.com>
wrote:

> *@dadah89* commented on this pull request.
> ------------------------------
>
> In lib/s3_meta_sync/syncer.rb
> <https://github.com/grosser/s3_meta_sync/pull/18#discussion_r132317887>:
>
> > @@ -83,6 +83,18 @@ def download(source, destination)
>        end
>      end
>
> +    def mktmpdir
>
> I'm using the syntax suggested here https://ruby-doc.org/stdlib-2.
> 3.1/libdoc/tmpdir/rdoc/Dir.html (last comment block)
> would renaming the method look better?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/grosser/s3_meta_sync/pull/18#discussion_r132317887>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ38bLHgO18iIFDtaflkK-OE75qBpks5sWis7gaJpZM4Oyock>
> .
>

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -83,6 +87,13 @@ def download(source, destination)
+    # Sometimes SIGTERM causes Dir.mktmpdir to not properly clear the temp folder
+    def clear_old_temp_folders
+      path = File.join(Dir.tmpdir, TEMP_FOLDER_PREFIX + '*')
+      total = Dir.glob(path).map { |dir| FileUtils.rm_rf(dir) }.count
```
needs to do some age checking or multiple parallel (maybe parallel projects / downloads) runs will delete each others folders

dadah89 replied:
oh that's true, lemme figure that out

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -244,22 +255,27 @@ def download_content(path)
+      ssl_error_retries ||= 0
+      ssl_error_retries += 1
+      if ssl_error_retries == 1
+        log "SSL error downloading #{path}, retrying #{ssl_error_retries}/#{max_retries}"
```
this should be "/2" since we only try twice ?

grosser replied:
... or 'retrying once'

dadah89 replied:
we only retry once, so will just remove that part :)

grosser commented:
looks less scary :)

dadah89 commented:
@grosser updated, tests still look good

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -83,6 +87,23 @@ def download(source, destination)
+        FileUtils.rm_rf(dir) if older_than(dir, day)
+      end.compact.count
+
+      log "Removed #{total} old temp folder(s)" if total > 0
```
might be interesting to say which folders were removed since that indicates a previous error ... and could be useful to search the logs for what failed

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -83,6 +87,23 @@ def download(source, destination)
+      total = Dir.glob(path).map do |dir|
+        day = 60 * 60 * 24
+        FileUtils.rm_rf(dir) if older_than(dir, day)
+      end.compact.count
```
semi dangerous in case rm_rf returns nil and it is then ignored due to compact ... but should be good

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -68,7 +71,8 @@ def download(source, destination)
       if download.any? || delete.any?
-        Dir.mktmpdir do |staging_area|
+        Dir.mktmpdir(TEMP_FOLDER_PREFIX) do |staging_area|
+          log "Temp folder: #{staging_area}"
```
maybe call it `Staging area` here too since that is what we name it internally ... and hopefully in the docs too ...

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -83,6 +87,23 @@ def download(source, destination)
+      path = File.join(Dir.tmpdir, TEMP_FOLDER_PREFIX + '*')
+
+      total = Dir.glob(path).map do |dir|
+        day = 60 * 60 * 24
```
usually written as `24 * 60 * 60`

dadah89 commented:
@grosser made some changes to address your comments, how does it look?

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -59,6 +60,8 @@ def upload(source, destination)
     end
 
     def download(source, destination)
+      clear_old_temp_folders
```
maybe be more verbose `delete_old_temp_folders`

grosser replied:
adn consistent delete_old_staging_areas

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -68,7 +71,8 @@ def download(source, destination)
       if download.any? || delete.any?
-        Dir.mktmpdir do |staging_area|
+        Dir.mktmpdir(STAGING_AREA_PREFIX) do |staging_area|
+          log "Temp folder: #{staging_area}"
```
log "Staging area #{staging_area}"

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -83,6 +87,21 @@ def download(source, destination)
       end
     end
 
+    # Sometimes SIGTERM causes Dir.mktmpdir to not properly clear the temp folder
```
clear -> delete

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -83,6 +87,21 @@ def download(source, destination)
     end
 
+    # Sometimes SIGTERM causes Dir.mktmpdir to not properly clear the temp folder
+    # Remove day old folders
```
`remove 1 day old folders`

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -83,6 +87,21 @@ def download(source, destination)
+      log "Removed #{removed} old temp folder(s)" if removed.count > 0
+    end
+
+    def older_than(dir, seconds)
```
`older_than?`

grosser replied:
maybe just inline that logic ...

```
paths = Dir.glob ...
paths.select! { } # only stale ones
paths.each ... rm
```

dadah89 replied:
That looks better. `older_than?` I was keeping a method so I can mock it, but will try mocking File.ctime instead

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -83,6 +87,21 @@ def download(source, destination)
+      path = File.join(Dir.tmpdir, STAGING_AREA_PREFIX + '*')
+
+      day = 24 * 60 * 60
+      removed = Dir.glob(path).select { |dir| older_than(dir, day) }. map { |dir| FileUtils.rm_rf(dir) }
```
`.flat_map` since rm_rf returns an array ... or just do `.each`

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -244,21 +263,26 @@ def download_content(path)
+      http_error_retries ||= 0
+      http_error_retries += 1
+      if http_error_retries <= max_retries
+        log "#{e.class} error downloading #{path}, retrying #{http_error_retries}/#{max_retries}"
```
inconsistent ... either show path or url or both in error message ... see `$!.message <<` below

grosser replied:
(I guess that's an old issue, but just saw it now)

grosser commented on lib/s3_meta_sync/syncer.rb:
```diff
@@ -244,21 +263,26 @@ def download_content(path)
+      with_retries(path: path, url: url) { open(url, options) }
+    end
+
+    def with_retries(path:, url:)
```
this is not a generic retry logic so maybe pick a more specific name `retry_downloads` or so

dadah89 commented:
@grosser I think I addressed all your comments, how does it look now?

tests:
```
(2.2.3) vagrant: /harmony/code/s3_meta_sync (igor/retry_remove_tmp_dir)$ rspec ./spec/s3_meta_sync_spec.rb
...........................................................

Finished in 1 minute 36.75 seconds (files took 0.25089 seconds to load)
59 examples, 0 failures

(2.4.1) vagrant: /harmony/code/s3_meta_sync (igor/retry_remove_tmp_dir)$ rspec ./spec/s3_meta_sync_spec.rb
...........................................................

Finished in 1 minute 40.24 seconds (files took 0.39314 seconds to load)
59 examples, 0 failures
```
